### PR TITLE
[CodeGen] Improve ExpandMemCmp for more efficient non-register aligned sizes handling

### PR DESCRIPTION
**Pull request by igogo-x86**
* Enhanced the logic of ExpandMemCmp pass to merge contiguous subsequences
  in LoadSequence, based on sizes allowed in `AllowedTailExpansions`.
* This enhancement seeks to minimize the number of basic blocks and produce
  optimized code when using memcmp with non-register aligned sizes.
* Enable this feature for AArch64 with memcmp sizes modulo 8 equal to
  3, 5, and 6.

### CODE REVIEW

File: llvm/lib/CodeGen/ExpandMemCmp.cpp [hunk 1]
```diff
@@ -279,6 +304,7 @@ void MemCmpExpansion::createResultBlock() {
 
 MemCmpExpansion::LoadPair MemCmpExpansion::getLoadPair(Type *LoadSizeType,
                                                        bool NeedsBSwap,
+                                                       Type *BSwapSizeType,
```
Comment (legrosbuffle) [inline llvm/lib/CodeGen/ExpandMemCmp.cpp:307]: We always have either (`NeedsBSwap == false`, `BSwapSizeType==nullptr`) or (`NeedsBSwap == true`, `BSwapSizeType!=nullptr`). We could simplify that.

File: llvm/lib/CodeGen/ExpandMemCmp.cpp [hunk 2]
```diff
@@ -255,6 +255,31 @@ MemCmpExpansion::MemCmpExpansion(
     }
   }
   assert(LoadSequence.size() <= Options.MaxNumLoads && "broken invariant");
+  // This part of code attempts to optimize the LoadSequence by merging allowed
+  // subsequences into single loads of allowed sizes from
+  // `AllowedTailExpansions`. If it is for zero comparison or if no allowed tail
+  // expansions are specified, we exit early.
+  if (IsUsedForZeroCmp || !Options.AllowedTailExpansions.size())
```
Comment (legrosbuffle) [inline llvm/lib/CodeGen/ExpandMemCmp.cpp:262]: `Options.AllowedTailExpansions.empty()`

File: llvm/lib/CodeGen/ExpandMemCmp.cpp [hunk 3]
```diff
@@ -255,6 +255,31 @@ MemCmpExpansion::MemCmpExpansion(
...
+    return;
+
+  while (LoadSequence.size() >= 2) {
+    auto Last = LoadSequence[LoadSequence.size() - 1];
+    auto PreLast = LoadSequence[LoadSequence.size() - 2];
+
```
Comment (legrosbuffle) [inline llvm/lib/CodeGen/ExpandMemCmp.cpp:268]: Can you move this to a separate function for consistency with the `computeOverlappingLoadSequence` code above ?

File: llvm/test/CodeGen/AArch64/memcmp.ll
Comment (legrosbuffle) [inline llvm/test/CodeGen/AArch64/memcmp.ll:1]: Can you submit the two test files as a baseline so that we can see the codegen diffs ?

Comment (david-arm) [reply llvm/test/CodeGen/AArch64/memcmp.ll:1]: Given github doesn't do support stacked pull requests like Phabricator did, I think @igogo-x86 tried to solve this problem by pushing two commits into this branch. The first commit shows the pre-committed tests, and the second commit does show the effect the code change has on the tests.

**Review by legrosbuffle — APPROVED** [review]

**Comment by DavidSpickett** [comment]
https://lab.llvm.org/buildbot/#/builders/185/builds/5239 is the plain AArch64 build, still in stage 1 at the moment.

**Comment by igogo-x86** [comment]
@DavidSpickett, yes, I reverted the commit. mecmp(..., ..., 3) on certain data was not correct. Going to investigate how to handle it properly

**Pull request merged**